esp32/mods/modpycom.c

```
@@ -162,8 +163,41 @@ STATIC mp_obj_t mod_pycom_pulses_get (mp_obj_t gpio, mp_obj_t timeout) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_2(mod_pycom_pulses_get_obj, mod_pycom_pulses_get);
 
+
+
+/*
+ * nvs_setstring/nvs_getstring: support for NVS string storage and, 
+ * thanks to JSON/other serializers, pretty much any datastructure.
+ */
+STATIC mp_obj_t mod_pycom_nvs_setstring (mp_obj_t _key, mp_obj_t _value) {
+    const char *key = mp_obj_str_get_str(_key);
+    const char *value = mp_obj_str_get_str(_value);
+    if (strlen(value) >= 1984) {
+        // orwell error: maximum length (including null character) is 1984 bytes
+        nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "value too long (max: 1984)"));
+        return mp_const_none;
+    }
+    esp_err_t esp_err = nvs_set_str(pycom_nvs_handle, key, value);
+    if (ESP_OK == esp_err) {
+        nvs_commit(pycom_nvs_handle);
+    } else if (ESP_ERR_NVS_NOT_ENOUGH_SPACE == esp_err || ESP_ERR_NVS_PAGE_FULL == esp_err || ESP_ERR_NVS_NO_FREE_PAGES == esp_err) {
+        nlr_raise(mp_obj_new_exception_msg(&mp_type_OSError, "no space available"));
+    } else if (ESP_ERR_NVS_INVALID_NAME == esp_err || ESP_ERR_NVS_KEY_TOO_LONG == esp_err) {
+        nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "invalid key (or too long)"));
+    } else {
+        // not ESP_OK, but not reporting?  TODO:check this...
```

Please also raise an exception here with a generic exception message